src/services/content.js

```diff
@@ -371,4 +373,320 @@ export async function getRecommendedForYou(brand, rowId = null, {
+        raw: p
+      }))
+  );
+}
+
+function mergeAndSortItems(items, limit) {
+  return items
+    .filter(item => typeof item.progressTimestamp === 'number' && item.progressTimestamp > 0)
+    .sort((a, b) => b.progressTimestamp - a.progressTimestamp)
+    .slice(0, limit + 5);
+}
+
+async function buildHierarchyMap(items) {
+  const contentIdsWithChildren = items
+    .filter(item => item.raw.child_count > 0)
+    .map(item => item.raw.id ?? item.raw.railcontent_id);
+
+  const entries = await Promise.all(
+    contentIdsWithChildren.map(async id => [id, await fetchHierarchy(id)])
+  );
+
+  return Object.fromEntries(entries);
+}
+
+async function buildChildContentMap(hierarchyMap) {
+  const allChildIds = Object.values(hierarchyMap).flatMap(h => Object.keys(h.children || {}));
+  const children = await fetchByRailContentIds(allChildIds);
+  return Object.fromEntries(children.map(child => [child.id, child]));
+}
+
+async function processPlaylistItem(item, progressMap) {
+  const playlist = item.raw;
+  const items = await fetchPlaylistItems(playlist.id);
+
+  const allItemsCompleted = items.every(i => {
+    const itemId = i.railcontent_id ?? i.id;
+    const progress = progressMap[itemId];
+    return progress && progress.status === 'completed';
+  });
+
+  if (allItemsCompleted) {
+    playlist.status = 'completed';
+    const firstItem = items[0];
+    if (firstItem) {
+      playlist.next = {
+        brand: playlist.brand,
+        id: playlist.id,
+        itemId: firstItem.id,
+        type: 'playlists',
+      };
+    }
+  } else {
+    const lastItemProgress = progressMap[playlist.last_engaged_on];
+
+    if (lastItemProgress?.status === 'completed') {
+      playlist.status = 'item_complete';
+      const index = items.findIndex(i => (i.railcontent_id ?? i.id) === playlist.last_engaged_on);
+      const nextItem = items[index + 1];
+      if (nextItem) {
+        playlist.next = {
+          brand: playlist.brand,
+          id: playlist.id,
+          itemId: nextItem.id,
+          type: 'playlists',
+        };
+      }
+    } else if (lastItemProgress?.status === 'started') {
+      playlist.status = 'item_started';
+      const currentItem = items.find(i => (i.railcontent_id ?? i.id) === playlist.last_engaged_on);
+      if (currentItem) {
+        playlist.next = {
+          brand: playlist.brand,
+          id: playlist.id,
+          itemId: currentItem.id,
+          type: 'playlists',
+        };
+      }
+    }
+  }
+
+  return {
+    id: playlist.id,
+    progressType: 'playlist',
+    header: 'playlist',
+    body: {
+      first_items_thumbnail_url: playlist.first_items_thumbnail_url,
+      title: playlist.name,
+      subtitle: `${playlist.duration_formated} • ${playlist.total_items} items • ${playlist.likes} likes • ${playlist.user.display_name}`
+    },
+    progressTimestamp: item.progressTimestamp,
+    cta: {
+      text: 'Continue',
+      action: playlist.next
+    }
```

Comment:
Some cleanup on this to reduce duplication and dead code.

```suggestion
  let nextId = items[0]?.id ?? null;
  if (!allItemsCompleted) {
    const lastItemProgress = progressMap[playlist.last_engaged_on];
    const index = items.findIndex(i => (i.railcontent_id ?? i.id) === playlist.last_engaged_on);
    if (lastItemProgress?.status === 'completed') {
      nextId = items[index + 1] ?? nextId;
    } else if (lastItemProgress?.status === 'started') {
      nextId = items[index] ?? nextId;
    }
  }

  return {
    id: playlist.id,
    progressType: 'playlist',
    header: 'playlist',
    body: {
      first_items_thumbnail_url: playlist.first_items_thumbnail_url,
      title: playlist.name,
      subtitle: `${playlist.duration_formated} • ${playlist.total_items} items • ${playlist.likes} likes • ${playlist.user.display_name}`
    },
    progressTimestamp: item.progressTimestamp,
    cta: {
      text: 'Continue',
      action: {
          brand: playlist.brand,
          id: playlist.id,
          itemId: nextItem.id,
          type: 'playlists',
        };
    }
```